src/cargo/ops/cargo_compile/unit_generator.rs

```diff
@@ -372,21 +372,6 @@ impl<'a> UnitGenerator<'a, '_> {
                             libs.push(proposal)
                         }
                     }
-                    if !all_targets && libs.is_empty() && *lib == LibRule::True {
```

Comment:
Seems the earliest ancestor of this error is this: https://github.com/rust-lang/cargo/pull/839/files#r1658124748